runtime/router.go

```diff
@@ -250,74 +150,28 @@ func (router *HTTPRouter) Register(
 	)
 }
 
-func (router *HTTPRouter) handleNotFound(w http.ResponseWriter, r *http.Request) {
-	resFields := []zapcore.Field{
-		zap.Int(statusCodeZapName, 404),
-	}
-	writeLogs(router.gateway.Logger, logRequestFields(r), resFields)
+func (router *HTTPRouter) handleNotFound(
+	w http.ResponseWriter,
+	r *http.Request,
+) {
+	req := NewServerHTTPRequest(w, r, nil, router.notFoundEndpoint)
 	// TODO custom NotFound
-	// A NotFound request is not started...
-	// TODO: inc.finish()
 	http.NotFound(w, r)
+	req.res.writeHeader(http.StatusNotFound)
+	req.res.finish()
 }
 
 func (router *HTTPRouter) handleMethodNotAllowed(
-	w http.ResponseWriter, r *http.Request,
+	w http.ResponseWriter,
+	r *http.Request,
 ) {
-	resFields := []zapcore.Field{
-		zap.Int(statusCodeZapName, 405),
-	}
-	writeLogs(router.gateway.Logger, logRequestFields(r), resFields)
+	req := NewServerHTTPRequest(w, r, nil, router.methodNotAllowedEndpoint)
 	// TODO: Remove coverage ignore when body unmarshaling supported.
 	// TODO custom MethodNotAllowed
 	http.Error(w,
 		http.StatusText(http.StatusMethodNotAllowed),
 		http.StatusMethodNotAllowed,
 	)
-}
-
-func logRequestFields(r *http.Request) []zapcore.Field {
-	// TODO: Allocating a fixed size array causes the zap logger to fail
-	// with ``unknown field type: { 0 0  <nil>}'' errors. Investigate this
-	// further to see if we can avoid reallocating underlying arrays for slices.
-	var fields []zapcore.Field
-	for k, v := range r.Header {
-		if len(v) > 0 {
-			fields = append(fields, zap.String("Request-Header-"+k, v[0]))
-		}
-	}
-
-	fields = append(fields, zap.String("method", r.Method))
-	fields = append(fields, zap.String("remoteAddr", r.RemoteAddr))
-	fields = append(fields, zap.String("pathname", r.URL.RequestURI()))
-	fields = append(fields, zap.String("host", r.Host))
-	fields = append(fields, zap.Time("timestamp", time.Now().UTC()))
-	// TODO add endpoint.id and endpoint.handlerId
-	// TODO log jaeger trace span
-
-	return fields
-}
-
-func logResponseFields(res *ServerHTTPResponse) []zapcore.Field {
-	var fields []zapcore.Field
-	fields = append(fields, zap.Int(statusCodeZapName, res.pendingStatusCode))
-	// TODO: Do not log body by default because PII and bandwidth.
-	// Temporarily log during the developement cycle
-	// TODO: Add a gateway level configurable body unmarshaller
-	// to extract only non-PII info.
-	fields = append(fields, zap.ByteString("Request Body", res.Request.RawBody))
-	fields = append(fields, zap.ByteString("Response Body", res.pendingBodyBytes))
-	fields = append(fields, zap.Time("timestamp-finished", res.finishTime))
-	return fields
-}
-
-func writeLogs(l *zap.Logger, reqFlds []zapcore.Field, resFlds []zapcore.Field) {
-	fields := reqFlds
-	if resFlds != nil {
-		fields = append(reqFlds, resFlds...)
-	}
-	l.Info(
-		"Finished an incoming server HTTP request",
-		fields...,
-	)
+	req.res.writeHeader(http.StatusMethodNotAllowed)
```

Comment:
why do we have to call `writeHeader` ? and can we do so before `http.Error()` ?

Reply:
Just need to set `req.res.StatusCode`, not call `writeHeader`